grails-app/controllers/org/pih/warehouse/inventory/InventoryController.groovy

```diff
@@ -515,15 +515,28 @@ class InventoryController {
         def location = Location.get(session.warehouse.id)
         def quantityMap = dashboardService.getInStock(location)
         def statusMap = dashboardService.getInventoryStatus(location)
+        def availableItems = []
+
         if (params.format == "csv") {
             def filename = "In stock - " + location.name + ".csv"
             response.setHeader("Content-disposition", "attachment; filename=\"${filename}\"")
             render(contentType: "text/csv", text: getCsvForProductMap(quantityMap, statusMap))
             return
+        } else {
+            quantityMap.each { Product product, quantity ->
+                def inventoryLevel = product.getInventoryLevel(session.warehouse.id)
+                def quantityAvailableToPromise = inventoryService.getQuantityAvailableToPromise(product, location)
```

Comment:
This is still doing a query per product. I was saying that we no longer need to get the quantityMap (above). Also, we could probably replace the query for inventoryLevel by doing a join in the getQuantityAvailableToPromise() query.
```
select <fields>
from product_availability
left outer join inventory level on product and location
```
And use the resultTransformer to return AvailableItems, including the inventoryLevel.

Reply:
Just make sure you only get one inventory level per product and location (i.e. you might need to join a subquery in order to ensure there's only one row per product, location returned by the inventory_level. It might look something like this in SQL.
```
select <fields>
from product_availability
left outer join (
    select 
        location_id, 
        product_id, 
        ... all the rest of the fields needed
    from inventory_level 
    join location on location.inventory_id = inventory_level.id
    group by location_id, product_id
    limit 1) as tmp on tmp.location_id = product_availability.location_id and tmp.product_id = product_availability.product_id
```